### DISCUSSION TITLE
Dedupe Identical Records & Similar Records Contextually Referring To The Same Thing

### DISCUSSION THREAD

**Discussion by NathanHGayle** [moj-analytical-services/splink#1565] [General]
#Goal
To dedupe a premises dataset with exact matching duplicates (discovered when combining all columns to create a unique_id and finding matchings in this Id field) and to dedupe records that refer to the same premises (not to be confused with field called prem) For example, two records have identical field values except their values for Address Line 1 has one extra/missing word like Bills bar vs Bills. This record is highly likely to be the same premises.

**Data** 
- 9100 records
- I've anonymized 5 records and attached here for context: (Prem here refers to a descriptive category not the actual premises)

[ExampleData.csv](https://github.com/moj-analytical-services/splink/files/12499377/ExampleData.csv)

**Expectations post EDA**

- I expect fewer duplicate values in combined address fields like "Address Line 1 + Address Line 2 + Postcode" since that would not be possible in reality (I should make this combination a column, thinking about it now),
- I do expect many duplicates in Number, Name, Prem, City and Local. So it's records with either everything identical or everything identical bar one generally unlikley difference like Addressline 1. 

**Output Desire**
- A dataframe column added with the probabilities that I can group to identify the duplicates

#Current code
` import pandas as pd
import splink
from splink.duckdb.duckdb_linker import DuckDBLinker
import splink.duckdb.duckdb_comparison_library as cl
print('Running Splink version:', splink.__version__)
df = pd.read_csv([ExampleData])  -- here I import the real dataset
settings = {
    "link_type": "dedupe_only",
    "blocking_rules_to_generate_predictions": [
        -- "l.Number = r.Number",
        --"l.Postcode = r.Postcode",
        --"l.Local = r.Local",
        --"l.City = r.City",
    ],
    "comparisons": [
        cl.exact_match("Number"),
        cl.levenshtein_at_thresholds("Name", [1,2],include_exact_match_level=False),
        cl.levenshtein_at_thresholds("Prem",[1,2],include_exact_match_level=False),
        cl.exact_match("Local"),
        cl.levenshtein_at_thresholds("Address_Line_1",[1,2],include_exact_match_level=False),
        cl.levenshtein_at_thresholds("Address_Line_2",[1,2],include_exact_match_level=False),
        cl.exact_match("City"),
        cl.exact_match("Postcode"),
        cl.exact_match("Postcode_District"),
    ],
}

linker = DuckDBLinker(df, settings)

linker.load_settings(settings) 

linker.estimate_u_using_random_sampling(1e8)
linker.save_settings_to_json("u_from_random_sample.json", overwrite=True)

training_session_an = linker.estimate_parameters_using_expectation_maximisation("l.city = r.city")
linker.save_settings_to_json("expectation_maximisation.json", overwrite=True)

e = linker.predict(threshold_match_probability= 0.2)

df_e = e.as_pandas_dataframe()

print(df_e)
`

#Error message

----- Estimating u probabilities using random sampling -----
u probability not trained for Prem - Levenshtein <= 2 (comparison vector value: 1). This usually means the comparison level was never observed in the training data.

Estimated u probabilities using random sampling

Your model is not yet fully trained. Missing estimates for:
    - Number (no m values are trained).
    - Name (no m values are trained).
    - Prem (some u values are not trained, no m values are trained).
    - Local (no m values are trained).
    - Address_Line_1 (no m values are trained).
    - Address_Line_2 (no m values are trained).
    - City (no m values are trained).
    - Postcode (no m values are trained).
    - Postcode_District (no m values are trained).

AND

Estimating the m probabilities of the model by blocking on:
l.city = r.city

Parameter estimates will be made for the following comparison(s):
    - Number
    - Name
    - Prem
    - Local
    - Address_Line_1
    - Address_Line_2
    - City
    - Postcode
    - Postcode_District

Parameter estimates cannot be made for the following comparison(s) since they are used in the blocking rules: 

WARNING:
Level Levenshtein <= 2 on comparison Prem not observed in dataset, unable to train m value

WARNING:
Level All other comparisons on comparison City not observed in dataset, unable to train m value
Iteration 1: Largest change in params was -0.919 in the m_probability of Address_Line_2, level `Levenshtein <= 1`

WARNING:
Level Levenshtein <= 2 on comparison Prem not observed in dataset, unable to train m value

WARNING:
Level All other comparisons on comparison City not observed in dataset, unable to train m value
Iteration 2: Largest change in params was 0.452 in the m_probability of Local, level `All other comparisons`

WARNING:
Level Levenshtein <= 2 on comparison Prem not observed in dataset, unable to train m value

WARNING:
Level All other comparisons on comparison City not observed in dataset, unable to train m value
Iteration 3: Largest change in params was 0.354 in the m_probability of Address_Line_1, level `All other comparisons`

WARNING:
Level Levenshtein <= 2 on comparison Prem not observed in dataset, unable to train m value

WARNING:
Level All other comparisons on comparison City not observed in dataset, unable to train m value
Iteration 4: Largest change in params was 0.398 in probability_two_random_records_match

WARNING:
Level Levenshtein <= 2 on comparison Prem not observed in dataset, unable to train m value

WARNING:
Level All other comparisons on comparison City not observed in dataset, unable to train m value
Iteration 5: Largest change in params was 0.0278 in probability_two_random_records_match

WARNING:
Level Levenshtein <= 2 on comparison Prem not observed in dataset, unable to train m value

WARNING:
Level All other comparisons on comparison City not observed in dataset, unable to train m value
Iteration 6: Largest change in params was 0.000482 in probability_two_random_records_match

WARNING:
Level Levenshtein <= 2 on comparison Prem not observed in dataset, unable to train m value

WARNING:
Level All other comparisons on comparison City not observed in dataset, unable to train m value
Iteration 7: Largest change in params was 7.93e-06 in probability_two_random_records_match

EM converged after 7 iterations
m probability not trained for Prem - Levenshtein <= 2 (comparison vector value: 1). This usually means the comparison level was never observed in the training data.
m probability not trained for City - All other comparisons (comparison vector value: 0). This usually means the comparison level was never observed in the training data.

Your model is not yet fully trained. Missing estimates for:
    - Prem (some u values are not trained, some m values are not trained).
    - City (some m values are not trained).

 -- WARNING --
You have called predict(), but there are some parameter estimates which have neither been estimated or specified in your settings dictionary.  To produce predictions the following untrained trained parameters will use default values.
Comparison: 'Prem':
    m values not fully trained
Comparison: 'Prem':
    u values not fully trained
Comparison: 'City':
    m values not fully trained

splink.exceptions.SplinkException: Error executing the following sql for table `__splink__truth_space_table`(__splink__truth_space_table_62f474fcb):

**Questions**:

1. Please can I have insight to the error meanings - I've summerized them to give you all the generaly gist 
2. How do I go about creating blocking rules that meet my context and output desire points above? am I going in the right direction here?
 

**What I've tried**

1) I have played with the "blocking_rules_to_generate_predictions" and as you can see I have even removed them entirely to see the changes in the error messages. 

2) The .as_pandas_dataframe() isn't firing but  pressume this is du to previous errors though it looks like a Duckdb error (when fully displayed) 

All help appreciated

**Comment by NathanHGayle** [comment]
Hi All,
Solved this by blocking on City only, then using the estimate_parameters_using_expectation_maximisation (l.City = r.City and l.Name = r.Name") to train M values followed by the estimate_u_using_random_sampling to train U values. 

Produced a 100,460 row dataset with great catches.

**Reply by RobinL** [reply]
Thanks for the detailed question and apologies for not responding sooner.  Glad you've got it working

You might also want to run a second EM training session on e.g.

`estimate_parameters_using_expectation_maximisation (l.postcode = r.postcode") `

which may improve your parameter values (and therefore accuracy) a little

**Reply by NathanHGayle** [reply]
Thanks @RobinL I've added this and it has improved  (based on the info from the waterfall chart insight)
